lambda/matching.ml

```diff
@@ -2047,15 +2044,15 @@ let make_record_matching p loc all_labels def = function
           in
           (access, str) :: make_args (pos + 1)
       in
-      let p = expand_record p in
-      let def = Default_environment.specialize (matcher_of_pattern p) def in
+      let head = expand_record_head head in
+      let def = Default_environment.specialize head def in
```

Comment:
Here, `expand_record_head` is called twice on the head: in the line above and in the call to matcher in `specialize`. Currently, this first call is necessary to ensure that we compute the right arity. But maybe, it is the definition of arity that should be updated to always compute the arity of expanded records?

Reply:
I'm not fond of your proposal for `arity`, it is too magic. I think that expanding twice is okay from the point of view of evolving the existing code. (It was not so clear about expansion from the beginning, we kept expansions around as we were not confident about what the global invariants were, and the code remains irregular as a result.) I'm a bit wary about doing more subtle things because the risk of regression (missing an expansion) seems worse than the gain as long as we remain in this style.
I tried to improve the code in a non-invasive way to push expansions around, in the context of this commit, and I got stuck into an ocean of not-quite-satisfying choices (including an attempt to rewrite `expand_record_head` without calling `expand_record` in a convoluted way; the attempt ended up being a disaster).

This question convinced me that your earlier proposition about tracking "head descriptions" in the code is the way forward: now I want to experiment with a private type that is equal to head desc, but always expanded (and maybe never `Any`?). This sort of type-directed discipline is the only way to get a good control of what's going on with expansion in the code, I think, but it would also be a fairly invasive change (touching a lot of different places).

I think that this particular commit should be left in peace (I see no simple way to avoid the double expansion, which is not a bug but merely a code smell). I don't know anymore whether we should try to add typing discipline for expansion in this PR, or in a PR right after that, or at the end of the PR stack.

Reply:
I think that a comment would be fine for now, to keep in mind that there is some duplication going one here.

Reply:
Done, thanks.